### PR TITLE
Memory

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/leonardoalt/yools/issues/18

### CODE REVIEW

File: tests/assert_pass/branches_memory.yul [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    let x := address()
+    mstore(0, 20)
+    mstore(32, 21)
+    if lt(x, 2) { mstore(0, 7) }
+    if gt(x, 1) { mstore(0, 8) }
+    let r := lt(mload(0), 9)
+
+    assert(eq(r, 1))
```
Comment (chriseth) [inline tests/assert_pass/branches_memory.yul:9]: I'm impressed that this is working.

File: src/evm_context.rs [hunk 2]
```diff
@@ -175,6 +211,13 @@ fn calldata() -> Identifier {
     identifier("_calldata", 1026)
 }
 
+fn memory() -> Identifier {
+    identifier("_memory", 1027)
+}
+fn memory_type() -> &'static str {
+    "(Array (_ BitVec 256) (_ BitVec 8))"
```
Comment (leonardoalt) [inline src/evm_context.rs:218]: We should probably extract these into a trait and have modules for memory/storage/calldata at some point.

File: src/evm_context.rs [hunk 3]
```diff
@@ -132,6 +141,33 @@ pub fn calldataload(index: &String, _ssa: &mut SSATracker) -> String {
...
 }
 
+pub fn mstore(index: &String, value: &String, ssa: &mut SSATracker) -> String {
+    let before = ssa.to_smt_variable(&memory());
+    let after = ssa.allocate_new_ssa_index(&memory());
+
+    let stored = (0..32).fold(before.name, |acc, i| {
+        let sub_index = format!("(bvadd {} #x{:064X})", index, i);
+        let byte_value = format!(
+            "((_ extract {} {}) {})",
```
Comment (leonardoalt) [inline src/evm_context.rs:151]: Why do you need to extract? Aren't the memory elements BV8 anyway?

File: src/ssa_tracker.rs [hunk 4]
```diff
@@ -11,6 +11,8 @@ pub struct SSATracker {
     /// TODO "highest" and "names" could be shared by copies of SSATracker...
     highest: BTreeMap<u64, u64>,
     names: BTreeMap<u64, String>,
+    /// Types of variables that are NOT 256 bit bitvectors.
+    types: HashMap<u64, String>,
```
Comment (leonardoalt) [inline src/ssa_tracker.rs:15]: Why HashMap instead of BTreeMap?

File: src/ssa_tracker.rs [hunk 5]
```diff
@@ -46,9 +48,11 @@ impl SSATracker {
                 let skipped_idx = self.current[key];
                 if skipped_idx != *value {
                     let new_ssa = self.allocate_new_ssa_index_by_id(*key);
+                    // TODO This does not work for memory
```
Comment (leonardoalt) [inline src/ssa_tracker.rs:51]: Why not?

File: src/ssa_tracker.rs [hunk 6]
```diff
@@ -46,9 +48,11 @@ impl SSATracker {
...
                     Some(format!(
-                        "(define-const {} (_ BitVec 256) (ite {} {} {}))\n",
+                        "(define-const {} {} (ite {} {} {}))\n",
                         self.id_to_smt_variable(*key, new_ssa).name,
+                        self.type_of_id(*key),
```
Comment (leonardoalt) [inline src/ssa_tracker.rs:55]: Isn't this returning only BV8 for now anyway?

File: src/evm_context.rs (hunk 3 above)
Comment (chriseth) [reply src/evm_context.rs:151]: No, this is the value to be stored, it's a regular stack word.

File: src/ssa_tracker.rs (hunk 4 above)
Comment (chriseth) [reply src/ssa_tracker.rs:15]: Because it's faster (is it?) and we don't need the ordering.

File: src/ssa_tracker.rs (hunk 5 above)
Comment (chriseth) [reply src/ssa_tracker.rs:51]: This todo is from before I introduced the types ;)

File: src/ssa_tracker.rs (hunk 6 above)
Comment (chriseth) [reply src/ssa_tracker.rs:55]: It's BV256 for all variables except the memory variable for which it returns an array type

Comment (leonardoalt) [reply src/ssa_tracker.rs:55]: where does that happen?

File: src/evm_context.rs (hunk 3 above)
Comment (leonardoalt) [reply src/evm_context.rs:151]: ah true

File: src/ssa_tracker.rs (hunk 6 above)
Comment (chriseth) [reply src/ssa_tracker.rs:55]: https://github.com/leonardoalt/yools/pull/66/files#diff-ab01308940caaee595a2da5f1569b16a0fd60fa02c7cb18e5d8d71b4e0918da8R27

File: src/ssa_tracker.rs (hunk 5 above)
Comment (leonardoalt) [reply src/ssa_tracker.rs:51]: So should it be removed now?

Comment (leonardoalt) [reply src/ssa_tracker.rs:51]: @chriseth

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**